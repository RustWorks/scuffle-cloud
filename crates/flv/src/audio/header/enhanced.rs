use std::io::{self, Read};

use byteorder::{BigEndian, ReadBytesExt};
use bytes::Bytes;
use nutype_enum::nutype_enum;
use scuffle_bytes_util::BytesCursorExt;

use crate::common::AvMultitrackType;
use crate::error::Error;

nutype_enum! {
    pub enum AudioPacketType(u8) {
        SeguenceStart = 0,
        CodedFrames = 1,
        SeguenceEnd = 2,
        MultichannelConfig = 4,
        Multitrack = 5,
        ModEx = 7,
    }
}

nutype_enum! {
    pub enum AudioPacketModExType(u8) {
        TimestampOffsetNano = 0,
    }
}

/// This is a helper enum to represent the different types of audio packet modifier extensions.
#[derive(Debug, Clone, PartialEq)]
pub enum AudioPacketModEx {
    TimestampOffsetNano {
        audio_timestamp_nano_offset: u32,
    },
    Other {
        audio_packet_mod_ex_type: AudioPacketModExType,
        mod_ex_data: Bytes,
    },
}

nutype_enum! {
    pub enum AudioFourCc([u8; 4]) {
        Ac3 = *b"ac-3",
        Eac3 = *b"ec-3",
        Opus = *b"Opus",
        Mp3 = *b".mp3",
        Flac = *b"fLaC",
        Aac = *b"mp4a",
    }
}

/// This is a helper enum to represent the different types of multitrack audio.
#[derive(Debug, Clone, PartialEq)]
pub enum ExAudioTagHeaderContent {
    NoMultiTrack(AudioFourCc),
    OneTrack(AudioFourCc),
    ManyTracks(AudioFourCc),
    ManyTracksManyCodecs,
    Other {
        audio_multitrack_type: AvMultitrackType,
        audio_four_cc: AudioFourCc,
    },
}

/// FLV `ExAudioTagHeader`
///
/// Defined by Enhanced RTMP v2 (Enhanced Audio section).
#[derive(Debug, Clone, PartialEq)]
pub struct ExAudioTagHeader {
    pub audio_packet_mod_exs: Vec<AudioPacketModEx>,
    pub audio_packet_type: AudioPacketType,
    pub content: ExAudioTagHeaderContent,
}

impl ExAudioTagHeader {
    pub fn demux(reader: &mut io::Cursor<Bytes>) -> Result<Self, Error> {
        let mut audio_packet_type = AudioPacketType::from(reader.read_u8()? & 0b0000_1111);

        let mut audio_packet_mod_exs = Vec::new();

        while audio_packet_type == AudioPacketType::ModEx {
            let mut mod_ex_data_size = reader.read_u8()? as usize + 1;
            if mod_ex_data_size == 256 {
                mod_ex_data_size = reader.read_u16::<BigEndian>()? as usize + 1;
            }

            let mod_ex_data = reader.extract_bytes(mod_ex_data_size)?;

            let next_byte = reader.read_u8()?;
            let audio_packet_mod_ex_type = AudioPacketModExType::from(next_byte >> 4); // 0b1111_0000
            audio_packet_type = AudioPacketType::from(next_byte & 0b0000_1111);

            if audio_packet_mod_ex_type == AudioPacketModExType::TimestampOffsetNano {
                if mod_ex_data_size < 3 {
                    // too few data bytes for the timestamp offset
                    return Err(Error::InvalidModExData { expected_bytes: 3 });
                }

                let mod_ex_data = &mut io::Cursor::new(mod_ex_data);
                audio_packet_mod_exs.push(AudioPacketModEx::TimestampOffsetNano {
                    audio_timestamp_nano_offset: mod_ex_data.read_u24::<BigEndian>()?,
                });
            } else {
                audio_packet_mod_exs.push(AudioPacketModEx::Other {
                    audio_packet_mod_ex_type,
                    mod_ex_data,
                });
            }
        }

        if audio_packet_type == AudioPacketType::Multitrack {
            let byte = reader.read_u8()?;
            let audio_multitrack_type = AvMultitrackType::from(byte >> 4); // 0b1111_0000
            audio_packet_type = AudioPacketType::from(byte & 0b0000_1111);

            if audio_packet_type == AudioPacketType::Multitrack {
                // nested multitracks are not allowed
                return Err(Error::NestedMultitracks);
            }

            let mut audio_four_cc = [0; 4];
            // Only read the four cc if it's not ManyTracksManyCodecs
            if audio_multitrack_type != AvMultitrackType::ManyTracksManyCodecs {
                reader.read_exact(&mut audio_four_cc)?;
            }

            let content = match audio_multitrack_type {
                AvMultitrackType::OneTrack => ExAudioTagHeaderContent::OneTrack(AudioFourCc::from(audio_four_cc)),
                AvMultitrackType::ManyTracks => ExAudioTagHeaderContent::ManyTracks(AudioFourCc::from(audio_four_cc)),
                AvMultitrackType::ManyTracksManyCodecs => ExAudioTagHeaderContent::ManyTracksManyCodecs,
                _ => ExAudioTagHeaderContent::Other {
                    audio_multitrack_type,
                    audio_four_cc: AudioFourCc::from(audio_four_cc),
                },
            };

            Ok(Self {
                audio_packet_mod_exs,
                audio_packet_type,
                content,
            })
        } else {
            let mut audio_four_cc = [0; 4];
            reader.read_exact(&mut audio_four_cc)?;
            let audio_four_cc = AudioFourCc::from(audio_four_cc);

            Ok(Self {
                audio_packet_mod_exs,
                audio_packet_type,
                content: ExAudioTagHeaderContent::NoMultiTrack(audio_four_cc),
            })
        }
    }
}
