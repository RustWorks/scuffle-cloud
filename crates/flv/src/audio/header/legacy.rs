use std::io;

use byteorder::ReadBytesExt;
use bytes::Bytes;
use nutype_enum::nutype_enum;

use crate::error::Error;

nutype_enum! {
    /// FLV Sound Format
    ///
    /// Denotes the type of the underlying data packet
    ///
    /// Defined by:
    /// - video_file_format_spec_v10.pdf (Chapter 1 - The FLV File Format - Audio tags)
    /// - video_file_format_spec_v10_1.pdf (Annex E.4.2.1 - AUDIODATA)
    pub enum SoundFormat(u8) {
        /// Linear PCM, platform endian
        LinearPcmPlatformEndian = 0,
        /// ADPCM
        Adpcm = 1,
        /// MP3
        Mp3 = 2,
        /// Linear PCM, little endian
        LinearPcmLittleEndian = 3,
        /// Nellymoser 16Khz Mono
        Nellymoser16KhzMono = 4,
        /// Nellymoser 8Khz Mono
        Nellymoser8KhzMono = 5,
        /// Nellymoser
        Nellymoser = 6,
        /// G.711 A-Law logarithmic PCM
        G711ALaw = 7,
        /// G.711 Mu-Law logarithmic PCM
        G711MuLaw = 8,
        /// The `ExAudioTagHeader` is present
        ///
        /// Defined by: Enhanced RTMP v2 (Enhanced Audio section)
        ExHeader = 9,
        /// AAC
        Aac = 10,
        /// Speex
        Speex = 11,
        /// Mp3 8Khz
        Mp38Khz = 14,
        /// Device specific sound
        DeviceSpecificSound = 15,
    }
}

nutype_enum! {
    /// FLV Sound Rate
    ///
    /// Denotes the sampling rate of the audio data.
    ///
    /// Defined by:
    /// - video_file_format_spec_v10.pdf (Chapter 1 - The FLV File Format - Audio tags)
    /// - video_file_format_spec_v10_1.pdf (Annex E.4.2.1 - AUDIODATA)
    pub enum SoundRate(u8) {
        /// 5.5 KHz
        Hz5500 = 0,
        /// 11 KHz
        Hz11000 = 1,
        /// 22 KHz
        Hz22000 = 2,
        /// 44 KHz
        Hz44000 = 3,
    }
}

nutype_enum! {
    /// FLV Sound Size
    ///
    /// Denotes the size of each sample in the audio data.
    ///
    /// Defined by:
    /// - video_file_format_spec_v10.pdf (Chapter 1 - The FLV File Format - Audio tags)
    /// - video_file_format_spec_v10_1.pdf (Annex E.4.2.1 - AUDIODATA)
    pub enum SoundSize(u8) {
        /// 8 bit
        Bit8 = 0,
        /// 16 bit
        Bit16 = 1,
    }
}

nutype_enum! {
    /// FLV Sound Type
    ///
    /// Denotes the number of channels in the audio data.
    ///
    /// Defined by:
    /// - video_file_format_spec_v10.pdf (Chapter 1 - The FLV File Format - Audio tags)
    /// - video_file_format_spec_v10_1.pdf (Annex E.4.2.1 - AUDIODATA)
    pub enum SoundType(u8) {
        /// Mono
        Mono = 0,
        /// Stereo
        Stereo = 1,
    }
}

/// The legacy FLV `AudioTagHeader` as defined by the original spec.
///
/// Defined by video_file_format_spec_v10_1.pdf (Annex E.4.2.1 - AUDIODATA).
#[derive(Debug, Clone, PartialEq)]
pub struct LegacyAudioTagHeader {
    /// The sound format of the audio data. (4 bits)
    pub sound_format: SoundFormat,
    /// The sound rate of the audio data. (2 bits)
    pub sound_rate: SoundRate,
    /// The sound size of the audio data. (1 bit)
    pub sound_size: SoundSize,
    /// The sound type of the audio data. (1 bit)
    pub sound_type: SoundType,
}

impl LegacyAudioTagHeader {
    #[allow(clippy::unusual_byte_groupings)]
    pub fn demux(reader: &mut io::Cursor<Bytes>) -> Result<Self, Error> {
        let byte = reader.read_u8()?;

        // SoundFormat is the first 4 bits of the byte
        let sound_format = SoundFormat::from(byte >> 4); // 0b1111_00_0_0
        // SoundRate is the next 2 bits of the byte
        let sound_rate = SoundRate::from((byte & 0b0000_11_0_0) >> 2);
        // SoundSize is the next bit of the byte
        let sound_size = SoundSize::from((byte & 0b0000_00_1_0) >> 1);
        // SoundType is the last bit of the byte
        let sound_type = SoundType::from(byte & 0b0000_00_0_1);

        Ok(Self {
            sound_format,
            sound_rate,
            sound_size,
            sound_type,
        })
    }
}

#[cfg(test)]
#[cfg_attr(all(test, coverage_nightly), coverage(off))]
mod tests {
    use super::*;

    #[test]
    fn test_sound_format() {
        let cases = [
            (
                0x00,
                SoundFormat::LinearPcmPlatformEndian,
                "SoundFormat::LinearPcmPlatformEndian",
            ),
            (0x01, SoundFormat::Adpcm, "SoundFormat::Adpcm"),
            (0x02, SoundFormat::Mp3, "SoundFormat::Mp3"),
            (0x03, SoundFormat::LinearPcmLittleEndian, "SoundFormat::LinearPcmLittleEndian"),
            (0x04, SoundFormat::Nellymoser16KhzMono, "SoundFormat::Nellymoser16KhzMono"),
            (0x05, SoundFormat::Nellymoser8KhzMono, "SoundFormat::Nellymoser8KhzMono"),
            (0x06, SoundFormat::Nellymoser, "SoundFormat::Nellymoser"),
            (0x07, SoundFormat::G711ALaw, "SoundFormat::G711ALaw"),
            (0x08, SoundFormat::G711MuLaw, "SoundFormat::G711MuLaw"),
            (0x0A, SoundFormat::Aac, "SoundFormat::Aac"),
            (0x0B, SoundFormat::Speex, "SoundFormat::Speex"),
            (0x0E, SoundFormat::Mp38Khz, "SoundFormat::Mp38Khz"),
            (0x0F, SoundFormat::DeviceSpecificSound, "SoundFormat::DeviceSpecificSound"),
        ];

        for (value, expected, name) in cases {
            let sound_format = SoundFormat::from(value);
            assert_eq!(sound_format, expected);
            assert_eq!(format!("{:?}", sound_format), name);
        }
    }

    #[test]
    fn test_sound_rate() {
        let cases = [
            (0x00, SoundRate::Hz5500, "SoundRate::Hz5500"),
            (0x01, SoundRate::Hz11000, "SoundRate::Hz11000"),
            (0x02, SoundRate::Hz22000, "SoundRate::Hz22000"),
            (0x03, SoundRate::Hz44000, "SoundRate::Hz44000"),
        ];

        for (value, expected, name) in cases {
            let sound_rate = SoundRate::from(value);
            assert_eq!(sound_rate, expected);
            assert_eq!(format!("{:?}", sound_rate), name);
        }
    }

    #[test]
    fn test_sound_size() {
        let cases = [
            (0x00, SoundSize::Bit8, "SoundSize::Bit8"),
            (0x01, SoundSize::Bit16, "SoundSize::Bit16"),
        ];

        for (value, expected, name) in cases {
            let sound_size = SoundSize::from(value);
            assert_eq!(sound_size, expected);
            assert_eq!(format!("{:?}", sound_size), name);
        }
    }

    #[test]
    fn test_sound_type() {
        let cases = [
            (0x00, SoundType::Mono, "SoundType::Mono"),
            (0x01, SoundType::Stereo, "SoundType::Stereo"),
        ];

        for (value, expected, name) in cases {
            let sound_type = SoundType::from(value);
            assert_eq!(sound_type, expected);
            assert_eq!(format!("{:?}", sound_type), name);
        }
    }
}
