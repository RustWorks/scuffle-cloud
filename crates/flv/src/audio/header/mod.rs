use std::io::{self, Seek};

use byteorder::ReadBytesExt;
use bytes::Bytes;

use crate::error::Error;

mod enhanced;
mod legacy;

pub use enhanced::*;
pub use legacy::*;

/// This is a helper enum to represent the different types of audio tag headers.
#[derive(Debug, Clone, PartialEq)]
pub enum AudioTagHeader {
    Legacy(LegacyAudioTagHeader),
    Enhanced(ExAudioTagHeader),
}

impl AudioTagHeader {
    #[allow(clippy::unusual_byte_groupings)]
    pub fn demux(reader: &mut io::Cursor<Bytes>) -> Result<Self, Error> {
        let byte = reader.read_u8()?;
        let sound_format = SoundFormat::from((byte & 0b1111_00_0_0) >> 4);
        // Seek back one byte because we need the other half of the byte again
        reader.seek_relative(-1)?;

        if sound_format == SoundFormat::ExHeader {
            ExAudioTagHeader::demux(reader).map(AudioTagHeader::Enhanced)
        } else {
            LegacyAudioTagHeader::demux(reader).map(AudioTagHeader::Legacy)
        }
    }
}
