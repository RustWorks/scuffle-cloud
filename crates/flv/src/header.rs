use std::io;

use byteorder::{BigEndian, ReadBytesExt};
use bytes::Bytes;
use scuffle_bytes_util::BytesCursorExt;

use crate::error::Error;

/// The FLV header
/// Whenever a FLV file is read these are the first 9 bytes of the file.
///
/// Defined by:
/// - video_file_format_spec_v10.pdf (Chapter 1 - The FLV Header - Page 8)
/// - video_file_format_spec_v10_1.pdf (Annex E.2 - The FLV Header)
#[derive(Debug, Clone, PartialEq)]
pub struct FlvHeader {
    /// The version of the FLV file.
    pub version: u8,
    /// Whether the FLV file contains audio tags.
    pub is_audio_present: bool,
    /// Whether the FLV file contains video tags.
    pub is_video_present: bool,
    /// The extra data in the FLV header.
    ///
    /// Since the header provides a data offset, this is the remaining bytes after the DataOffset field
    /// to the end of the header.
    pub extra: Bytes,
}

impl FlvHeader {
    /// Demux the FLV header from the given reader.
    /// The reader will be returned in the position of the start of the data
    /// offset.
    pub fn demux(reader: &mut io::Cursor<Bytes>) -> Result<Self, Error> {
        let start = reader.position() as usize;

        let signature = reader.read_u24::<BigEndian>()?;

        // 0 byte at the beginning because we are only reading 3 bytes not 4.
        if signature != u32::from_be_bytes([0, b'F', b'L', b'V']) {
            return Err(Error::InvalidSignature(signature));
        }

        let version = reader.read_u8()?;
        let flags = reader.read_u8()?;
        let is_audio_present = (flags & 0b00000100) != 0;
        let is_video_present = (flags & 0b00000001) != 0;

        let data_offset = reader.read_u32::<BigEndian>()?;
        let end = reader.position() as usize;
        let size = end - start;

        let remaining = (data_offset as usize)
            .checked_sub(size)
            .ok_or(Error::InvalidDataOffset(data_offset))?;

        let extra = reader.extract_bytes(remaining)?;

        Ok(FlvHeader {
            version,
            is_audio_present,
            is_video_present,
            extra,
        })
    }
}
