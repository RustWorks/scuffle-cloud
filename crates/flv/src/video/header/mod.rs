use std::io::{self, Seek};

use byteorder::ReadBytesExt;
use bytes::Bytes;
use nutype_enum::nutype_enum;

use crate::error::Error;

pub mod enhanced;
pub mod legacy;

nutype_enum! {
    /// FLV Frame Type
    /// This enum represents the different types of frames in a FLV file.
    /// Defined by:
    /// - video_file_format_spec_v10.pdf (Chapter 1 - The FLV File Format - Video tags)
    /// - video_file_format_spec_v10_1.pdf (Annex E.4.3.1 - VIDEODATA)
    pub enum VideoFrameType(u8) {
        /// A keyframe is a frame that is a complete representation of the video content.
        KeyFrame = 1,
        /// An interframe is a frame that is a partial representation of the video content.
        InterFrame = 2,
        /// A disposable interframe is a frame that is a partial representation of the video content, but is not required to be displayed. (h263 only)
        DisposableInterFrame = 3,
        /// A generated keyframe is a frame that is a complete representation of the video content, but is not a keyframe. (reserved for server use only)
        GeneratedKeyFrame = 4,
        /// A video info or command frame is a frame that contains video information or commands.
        /// If the frame is this type, the body will be a CommandPacket
        Command = 5,
    }
}

nutype_enum! {
    pub enum VideoCommand(u8) {
        StartSeek = 0,
        EndSeek = 1,
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum VideoTagHeaderData {
    Legacy(legacy::LegacyVideoTagHeader),
    Enhanced(enhanced::ExVideoTagHeader),
}

#[derive(Debug, Clone, PartialEq)]
pub struct VideoTagHeader {
    /// The frame type of the video data. (4 bits)
    pub frame_type: VideoFrameType,
    pub data: VideoTagHeaderData,
}

impl VideoTagHeader {
    #[allow(clippy::unusual_byte_groupings)]
    pub fn demux(reader: &mut io::Cursor<Bytes>) -> Result<Self, Error> {
        let byte = reader.read_u8()?;
        // seek back one byte so that the codec id can be read again
        reader.seek_relative(-1)?;

        let is_ex_video_header = (byte & 0b1_000_0000) != 0;

        let data = if !is_ex_video_header {
            VideoTagHeaderData::Legacy(legacy::LegacyVideoTagHeader::demux(reader)?)
        } else {
            VideoTagHeaderData::Enhanced(enhanced::ExVideoTagHeader::demux(reader)?)
        };

        Ok(VideoTagHeader {
            frame_type: VideoFrameType::from((byte & 0b0_111_0000) >> 4),
            data,
        })
    }
}
