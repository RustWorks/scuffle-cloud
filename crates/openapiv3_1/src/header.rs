//! Implements [OpenAPI Header Object][header] types.
//!
//! [header]: https://spec.openapis.org/oas/latest.html#header-object

use super::{Object, Schema, Type};

/// Implements [OpenAPI Header Object][header] for response headers.
///
/// [header]: https://spec.openapis.org/oas/latest.html#header-object
#[non_exhaustive]
#[derive(serde_derive::Serialize, serde_derive::Deserialize, Clone, PartialEq, bon::Builder)]
#[cfg_attr(feature = "debug", derive(Debug))]
#[builder(on(_, into))]
pub struct Header {
    /// Schema of header type.
    pub schema: Schema,

    /// Additional description of the header value.
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub description: Option<String>,
}

impl Header {
    /// Construct a new [`Header`] with custom schema. If you wish to construct a default
    /// header with `String` type you can use [`Header::default`] function.
    ///
    /// # Examples
    ///
    /// Create new [`Header`] with integer type.
    /// ```rust
    /// # use openapiv3_1::header::Header;
    /// # use openapiv3_1::{Object, Type};
    /// let header = Header::new(Object::with_type(Type::Integer));
    /// ```
    ///
    /// Create a new [`Header`] with default type `String`
    /// ```rust
    /// # use openapiv3_1::header::Header;
    /// let header = Header::default();
    /// ```
    pub fn new<C: Into<Schema>>(component: C) -> Self {
        Self {
            schema: component.into(),
            ..Default::default()
        }
    }
}

impl Default for Header {
    fn default() -> Self {
        Self {
            description: Default::default(),
            schema: Object::builder().schema_type(Type::String).into(),
        }
    }
}

impl<S: header_builder::IsComplete> From<HeaderBuilder<S>> for Header {
    fn from(builder: HeaderBuilder<S>) -> Self {
        builder.build()
    }
}
