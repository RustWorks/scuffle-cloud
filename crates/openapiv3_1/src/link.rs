//! Implements [Open API Link Object][link_object] for responses.
//!
//! [link_object]: https://spec.openapis.org/oas/latest.html#link-object
use indexmap::IndexMap;
use serde_derive::{Deserialize, Serialize};

use super::Server;
use super::extensions::Extensions;

/// Implements [Open API Link Object][link_object] for responses.
///
/// The `Link` represents possible design time link for a response. It does not guarantee
/// callers ability to invoke it but rather provides known relationship between responses and
/// other operations.
///
/// For computing links, and providing instructions to execute them,
/// a runtime [expression][expression] is used for accessing values in an operation
/// and using them as parameters while invoking the linked operation.
///
/// [expression]: https://spec.openapis.org/oas/latest.html#runtime-expressions
/// [link_object]: https://spec.openapis.org/oas/latest.html#link-object
#[derive(Serialize, Deserialize, Clone, PartialEq, Default, bon::Builder)]
#[cfg_attr(feature = "debug", derive(Debug))]
#[builder(on(_, into))]
#[non_exhaustive]
pub struct Link {
    /// A map representing parameters to pass to an operation as specified with _`operation_id`_
    /// or identified by _`operation_ref`_. The key is parameter name to be used and value can
    /// be any value supported by JSON or an [expression][expression] e.g. `$path.id`
    ///
    /// [expression]: https://spec.openapis.org/oas/latest.html#runtime-expressions
    #[serde(skip_serializing_if = "IndexMap::is_empty", default)]
    #[builder(field)]
    pub parameters: IndexMap<String, serde_json::Value>,

    /// A relative or absolute URI reference to an OAS operation. This field is
    /// mutually exclusive of the _`operation_id`_ field, and **must** point to an [Operation
    /// Object][operation].
    /// Relative _`operation_ref`_ values may be used to locate an existing [Operation
    /// Object][operation] in the OpenAPI definition. See the rules for resolving [Relative
    /// References][relative_references].
    ///
    /// [relative_references]: https://spec.openapis.org/oas/latest.html#relative-references-in-uris
    /// [operation]: ../path/struct.Operation.html
    #[serde(skip_serializing_if = "String::is_empty", default)]
    #[builder(default)]
    pub operation_ref: String,

    /// The name of an existing, resolvable OAS operation, as defined with a unique
    /// _`operation_id`_.
    /// This field is mutually exclusive of the _`operation_ref`_ field.
    #[serde(skip_serializing_if = "String::is_empty", default)]
    #[builder(default)]
    pub operation_id: String,

    /// A literal value or an [expression][expression] to be used as request body when operation is called.
    ///
    /// [expression]: https://spec.openapis.org/oas/latest.html#runtime-expressions
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub request_body: Option<serde_json::Value>,

    /// Description of the link. Value supports Markdown syntax.
    #[serde(skip_serializing_if = "String::is_empty", default)]
    #[builder(default)]
    pub description: String,

    /// A [`Server`][server] object to be used by the target operation.
    ///
    /// [server]: ../server/struct.Server.html
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub server: Option<Server>,

    /// Optional extensions "x-something".
    #[serde(skip_serializing_if = "Option::is_none", default, flatten)]
    pub extensions: Option<Extensions>,
}

impl<S: link_builder::State> LinkBuilder<S> {
    /// Add parameters to be passed to [Operation][operation] upon execution.
    ///
    /// [operation]: ../path/struct.Operation.html
    pub fn parameters<N: Into<String>, V: Into<serde_json::Value>>(self, items: impl IntoIterator<Item = (N, V)>) -> Self {
        items.into_iter().fold(self, |this, (n, v)| this.parameter(n, v))
    }

    /// Add parameter to be passed to [Operation][operation] upon execution.
    ///
    /// [operation]: ../path/struct.Operation.html
    pub fn parameter<N: Into<String>, V: Into<serde_json::Value>>(mut self, name: N, value: V) -> Self {
        self.parameters.insert(name.into(), value.into());
        self
    }
}

impl<S: link_builder::IsComplete> From<LinkBuilder<S>> for Link {
    fn from(builder: LinkBuilder<S>) -> Self {
        builder.build()
    }
}
