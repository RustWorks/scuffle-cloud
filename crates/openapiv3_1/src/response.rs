//! Implements [OpenApi Responses][responses].
//!
//! [responses]: https://spec.openapis.org/oas/latest.html#responses-object
use indexmap::IndexMap;
use serde_derive::{Deserialize, Serialize};

use super::Content;
use super::extensions::Extensions;
use super::header::Header;
use super::link::Link;
use crate::{Ref, RefOr};

/// Implements [OpenAPI Responses Object][responses].
///
/// Responses is a map holding api operation responses identified by their status code.
///
/// [responses]: https://spec.openapis.org/oas/latest.html#responses-object
#[non_exhaustive]
#[derive(Serialize, Deserialize, Default, Clone, PartialEq, bon::Builder)]
#[cfg_attr(feature = "debug", derive(Debug))]
#[serde(rename_all = "camelCase")]
#[builder(on(_, into))]
pub struct Responses {
    /// Map containing status code as a key with represented response as a value.
    #[serde(flatten)]
    #[builder(field)]
    pub responses: IndexMap<String, RefOr<Response>>,

    /// Optional extensions "x-something".
    #[serde(skip_serializing_if = "Option::is_none", flatten)]
    pub extensions: Option<Extensions>,
}

impl Responses {
    /// Construct a new [`Responses`].
    pub fn new() -> Self {
        Default::default()
    }
}

impl<S: responses_builder::State> ResponsesBuilder<S> {
    /// Add a [`Response`].
    pub fn response(mut self, code: impl Into<String>, response: impl Into<RefOr<Response>>) -> Self {
        self.responses.insert(code.into(), response.into());

        self
    }

    /// Add responses from an iterator over a pair of `(status_code, response): (String, Response)`.
    pub fn responses_from_iter<I: IntoIterator<Item = (C, R)>, C: Into<String>, R: Into<RefOr<Response>>>(
        mut self,
        iter: I,
    ) -> Self {
        self.responses
            .extend(iter.into_iter().map(|(code, response)| (code.into(), response.into())));
        self
    }
}

impl From<Responses> for IndexMap<String, RefOr<Response>> {
    fn from(responses: Responses) -> Self {
        responses.responses
    }
}

impl<C, R> FromIterator<(C, R)> for Responses
where
    C: Into<String>,
    R: Into<RefOr<Response>>,
{
    fn from_iter<T: IntoIterator<Item = (C, R)>>(iter: T) -> Self {
        Self {
            responses: IndexMap::from_iter(iter.into_iter().map(|(code, response)| (code.into(), response.into()))),
            ..Default::default()
        }
    }
}

/// Implements [OpenAPI Response Object][response].
///
/// Response is api operation response.
///
/// [response]: https://spec.openapis.org/oas/latest.html#response-object
#[non_exhaustive]
#[derive(Serialize, Deserialize, Default, Clone, PartialEq, bon::Builder)]
#[cfg_attr(feature = "debug", derive(Debug))]
#[serde(rename_all = "camelCase")]
#[builder(on(_, into))]
pub struct Response {
    /// Map of headers identified by their name. `Content-Type` header will be ignored.
    #[serde(skip_serializing_if = "IndexMap::is_empty", default)]
    #[builder(field)]
    pub headers: IndexMap<String, Header>,

    /// Map of response [`Content`] objects identified by response body content type e.g `application/json`.
    ///
    /// [`Content`]s are stored within [`IndexMap`] to retain their insertion order. Swagger UI
    /// will create and show default example according to the first entry in `content` map.
    #[serde(skip_serializing_if = "IndexMap::is_empty", default)]
    #[builder(field)]
    pub content: IndexMap<String, Content>,

    /// A map of operations links that can be followed from the response. The key of the
    /// map is a short name for the link.
    #[serde(skip_serializing_if = "IndexMap::is_empty", default)]
    #[builder(field)]
    pub links: IndexMap<String, RefOr<Link>>,

    /// Optional extensions "x-something".
    #[serde(skip_serializing_if = "Option::is_none", flatten)]
    pub extensions: Option<Extensions>,

    /// Description of the response. Response support markdown syntax.
    pub description: String,
}

impl Response {
    /// Construct a new [`Response`].
    ///
    /// Function takes description as argument.
    pub fn new<S: Into<String>>(description: S) -> Self {
        Self {
            description: description.into(),
            ..Default::default()
        }
    }
}

impl<S: response_builder::State> ResponseBuilder<S> {
    /// Add [`Content`] of the [`Response`] with content type e.g `application/json`.
    pub fn content(mut self, content_type: impl Into<String>, content: impl Into<Content>) -> Self {
        self.content.insert(content_type.into(), content.into());
        self
    }

    /// Add response [`Header`].
    pub fn header(mut self, name: impl Into<String>, header: impl Into<Header>) -> Self {
        self.headers.insert(name.into(), header.into());
        self
    }

    /// Add link that can be followed from the response.
    pub fn link(mut self, name: impl Into<String>, link: impl Into<RefOr<Link>>) -> Self {
        self.links.insert(name.into(), link.into());
        self
    }

    /// Add [`Content`] of the [`Response`] with content type e.g `application/json`.
    pub fn contents<A: Into<String>, B: Into<Content>>(self, contents: impl IntoIterator<Item = (A, B)>) -> Self {
        contents.into_iter().fold(self, |this, (a, b)| this.content(a, b))
    }

    /// Add response [`Header`].
    pub fn headers<A: Into<String>, B: Into<Header>>(self, headers: impl IntoIterator<Item = (A, B)>) -> Self {
        headers.into_iter().fold(self, |this, (a, b)| this.header(a, b))
    }

    /// Add link that can be followed from the response.
    pub fn links<A: Into<String>, B: Into<RefOr<Link>>>(self, links: impl IntoIterator<Item = (A, B)>) -> Self {
        links.into_iter().fold(self, |this, (a, b)| this.link(a, b))
    }
}

impl<S: response_builder::IsComplete> From<ResponseBuilder<S>> for Response {
    fn from(builder: ResponseBuilder<S>) -> Self {
        builder.build()
    }
}

impl<S: response_builder::IsComplete> From<ResponseBuilder<S>> for RefOr<Response> {
    fn from(builder: ResponseBuilder<S>) -> Self {
        Self::T(builder.build())
    }
}

impl From<Ref> for RefOr<Response> {
    fn from(r: Ref) -> Self {
        Self::Ref(r)
    }
}

#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
mod tests {
    use insta::assert_json_snapshot;

    use super::{Content, Responses};
    use crate::Response;

    #[test]
    fn responses_new() {
        let responses = Responses::new();

        assert!(responses.responses.is_empty());
    }

    #[test]
    fn response_builder() {
        let request_body = Response::builder()
            .description("A sample response")
            .content(
                "application/json",
                Content::new(Some(crate::Ref::from_schema_name("MySchemaPayload"))),
            )
            .build();
        assert_json_snapshot!(request_body);
    }
}
