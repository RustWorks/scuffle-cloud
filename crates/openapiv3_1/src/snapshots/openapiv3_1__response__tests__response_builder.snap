---
source: crates/openapiv3_1/src/response.rs
expression: request_body
---
{
  "content": {
    "application/json": {
      "schema": {
        "$ref": "#/components/schemas/MySchemaPayload"
      }
    }
  },
  "description": "A sample response"
}
